src/theory/bv/theory_bv_solver.h

```diff
@@ -0,0 +1,130 @@
+/*********************                                                        */
+/*! \file theory_bv.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Liana Hadarean, Andrew Reynolds, Tim King
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Bitvector theory.
+ **
+ ** Bitvector theory.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__BV__THEORY_BV_SOLVER_H
+#define CVC4__THEORY__BV__THEORY_BV_SOLVER_H
+
+#include "theory/bv/theory_bv.h"
+#include "theory/theory.h"
+
+namespace CVC4 {
+namespace theory {
+namespace bv {
+
+class TheoryBVSolver
+{
+ public:
+  TheoryBVSolver(TheoryBV& bv) : d_bv(bv){};
+
+  virtual ~TheoryBVSolver(){};
+
+  /**
+   * Returns true if we need an equality engine. If so, we initialize the
+   * information regarding how it should be setup. For details, see the
+   * documentation in Theory::needsEqualityEngine.
+   */
+  virtual bool needsEqualityEngine(EeSetupInfo& esi) { return false; }
+
+  virtual void finishInit(){};
+
+  virtual void preRegisterTerm(TNode n) = 0;
+
+  virtual void check(Theory::Effort e) = 0;
```

Comment:
This method should not be overridden by new solvers.  Instead, use preCheck / postCheck / preNotifyFact / notifyFact (these should probably be defined as virtual methods here).

Reply:
Also notice that TheoryBVLazy can implement its current check as a preCheck, returning true.  (The new standard is strictly more flexible wrt the old one).